File: Algo/Algorithm/Questions/Arrays/21.calendarMatching/main.js
// O(c1 + c2) time | O(c1 + c2) space - where c1 and c2 are the respective numbers of meetings in calendar1 and calendar2.
function timeToMinutes(time) {
  const [hours, minutes] = time.split(':').map(str => parseInt(str, 10));
  return hours * 60 + minutes;
}
function updateCalendar(calendar, dailyBounds) {
  const updatedCalendar = [['0:00', dailyBounds[0]], ...calendar, [dailyBounds[1], '23:59']];

  return updatedCalendar.map(meeting => meeting.map(timeToMinutes));
}

function mergeCalendars(calendar1, calendar2) {
  const mergedCalendars = [];
  let i = 0;
  let j = 0;

  while (i < calendar1.length && j < calendar2.length) {
    if (calendar1[i][0] < calendar2[j][0]) {
      mergedCalendars.push(calendar1[i]);
      i += 1;
    } else {
      mergedCalendars.push(calendar2[j]);
      j += 1;
    }
  }

  while (i < calendar1.length) {
    mergedCalendars.push(calendar1[i]);
    i += 1;
  }

  while (j < calendar2.length) {
    mergedCalendars.push(calendar2[j]);
    j += 1;
  }

  return mergedCalendars;
}

function flattenCalendar(calendar) {
  const flattened = [calendar[0]];

  for (let i = 1; i < calendar.length; i += 1) {
    const currMeeting = calendar[i];
    const prevMeeting = flattened[flattened.length - 1];

    const [currStart, currEnd] = currMeeting;
    const [prevStart, prevEnd] = prevMeeting;

    if (prevEnd >= currStart) {
      const newCurrMeeting = [prevStart, Math.max(prevEnd, currEnd)];
      flattened[flattened.length - 1] = newCurrMeeting;
    } else {
      flattened.push(currMeeting);
    }
  }

  return flattened;
}

function minutesToTime(match) {
  const hour = Math.floor(match / 60);
  let minutes = match % 60;
  if (minutes < 10) {
    minutes = `0${minutes}`;
  }

  return `${hour}:${minutes}`;
}

function getMatchingAvailabilities(calendar, meetingDuration) {
  const matchingAvailabilities = [];

  for (let i = 1; i < calendar.length; i += 1) {
    const start = calendar[i - 1][1];
    const end = calendar[i][0];
    const availabilityDuration = end - start;

    if (availabilityDuration >= meetingDuration) {
      matchingAvailabilities.push([start, end]);
    }
  }

  return matchingAvailabilities.map(match => match.map(minutesToTime));
}

function calendarMatching(calendar1, dailyBounds1, calendar2, dailyBounds2, meetingDuration) {
  const updatedCalendar1 = updateCalendar(calendar1, dailyBounds1);
  const updatedCalendar2 = updateCalendar(calendar2, dailyBounds2);
  const mergedCalendars = mergeCalendars(updatedCalendar1, updatedCalendar2);
  const flattedCalendar = flattenCalendar(mergedCalendars);

  return getMatchingAvailabilities(flattedCalendar, meetingDuration);
}

File: Algo/Algorithm/Questions/Arrays/21.calendarMatching/main.cpp
#include <vector>
#include <iostream>
using namespace std;

struct StringMeeting
{
  string start;
  string end;
};

struct Meeting
{
  int start;
  int end;
};

int timeToMinutes(string time)
{
  int delimiter = time.find(":");
  int hour = stoi(time.substr(0, delimiter));
  int minutes = stoi(time.substr(delimiter + 1, time.length()));
  return hour * 60 + minutes;
}

vector<Meeting>
updateCalendar(
    vector<StringMeeting> calendar,
    StringMeeting dailyBounds)
{
  vector<StringMeeting> updatedCalendar;
  updatedCalendar.push_back({"0:00", dailyBounds.start});
  updatedCalendar.insert(updatedCalendar.end(), calendar.begin(), calendar.end());
  updatedCalendar.push_back({dailyBounds.end, "23:59"});
  vector<Meeting> calendarInMinutes;
  for (auto x : updatedCalendar)
    calendarInMinutes.push_back({timeToMinutes(x.start), timeToMinutes(x.end)});

  return calendarInMinutes;
}

vector<Meeting> findAvailableTime(
    vector<Meeting> calendar)
{
  vector<Meeting> avilableCalendar;
  for (int i = 0; i < (calendar.size() - 1); i++)
  {
    int startTime = calendar[i].end;
    int endTime = calendar[i + 1].start;
    avilableCalendar.push_back({startTime, endTime});
  }
  return avilableCalendar;
}

vector<Meeting> mergeAvailableClalendars(vector<Meeting> calendar1, vector<Meeting> calendar2)
{
  vector<Meeting> mergedCalendar;
  int i = 0;
  int j = 0;
  while (i < calendar1.size() && j < calendar2.size())
  {
    Meeting currCalendar1 = calendar1[i];
    Meeting currCalendar2 = calendar2[j];
    if (currCalendar1.start < currCalendar2.start)
    {
      mergedCalendar.push_back(currCalendar1);
      i++;
    }
    else
    {
      mergedCalendar.push_back(currCalendar2);
      j++;
    }
  }

  while (i < calendar1.size())
    mergedCalendar.push_back(calendar1[i++]);
  while (j < calendar2.size())
    mergedCalendar.push_back(calendar2[j++]);

  return mergedCalendar;
}

vector<Meeting> calendarMatching(vector<StringMeeting> calendar1,
                                 StringMeeting dailyBounds1,
                                 vector<StringMeeting> calendar2,
                                 StringMeeting dailyBounds2,
                                 int meetingDuration)
{
  // Write your code here.
  vector<Meeting> updatedCalendar1 = updateCalendar(calendar1, dailyBounds1);
  vector<Meeting> updatedCalendar2 = updateCalendar(calendar2, dailyBounds2);
  vector<Meeting> avilableCalendar1 = findAvailableTime(updatedCalendar1);
  vector<Meeting> avilableCalendar2 = findAvailableTime(updatedCalendar2);
  vector<Meeting> mergedCalendars = mergeAvailableClalendars(avilableCalendar1, avilableCalendar2);
  return mergedCalendars;
}

int main()
{
  vector<StringMeeting> calendar1 = {
      {"9:00", "10:30"},
      {"12:00", "13:00"},
      {"16:00", "18:00"}};
  vector<StringMeeting> calendar2 = {
      {"10:00", "11:30"},
      {"12:30", "14:30"},
      {"14:30", "15:00"},
      {"16:00", "17:00"}};
  StringMeeting dailyBounds1 = {"9:00", "20:00"};
  StringMeeting dailyBounds2 = {"9:00", "20:00"};
  int meetingDuration = 30;

  for (auto x : calendarMatching(calendar1, dailyBounds1, calendar2, dailyBounds2, meetingDuration))
    cout << x.start << " " << x.end << endl;
  return 0;
}